pennylane/transforms/get_unitary_matrix.py

```diff
@@ -0,0 +1,119 @@
+from difflib import SequenceMatcher
+from functools import wraps, reduce
+
+from pennylane.wires import Wires
+from pennylane.tape import QuantumTape, get_active_tape
+import numpy as np
+
+
+def get_unitary_matrix(fn, wire_order):
+    """Given a quantum circuit and a list of wire ordering, construct the matrix representation"""
+
+    wire_order = Wires(wire_order)
+    n_wires = len(wire_order)
+
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        ######### NOTE BELOW
+        active_tape = get_active_tape()
+
+        if active_tape is not None:
+            with active_tape.stop_recording(), QuantumTape() as tape:
+                fn(*args, **kwargs)
+        else:
+            # Not within a queuing context
+            with QuantumTape() as tape:
+                fn(*args, **kwargs)
+
+        if not tape.operations:
+            # we called op.expand(): get the outputted tape
+            tape = fn(*args, **kwargs)
+        #### NOTE took the above from qml.adjoint. Appropriate here?
```

Comment:
(We also have a [new blog post](https://pennylane.ai/blog/2021/08/how-to-write-quantum-function-transforms-in-pennylane/) that discusses this stuff more in depth.

Reply:
Aha! Thanks for the tips the blogpost certaintly helped understanding tapes. But I'm a bit unsure how to use `qfunc_transform`. Before, I interpreted "quantum function transform" as _transforms that acts on quantum functions_ as in the documentation for `qml.transforms`... Getting rid of the tape stuff would be nice, but I don't know how I'm expected to use `qfunc_transform` when I don't want to return another function, but a matrix (?)

Reply:
@ingstra hmm this is a very good point... now that you mention this, it doesn't fit the current framework of `qfunc_transform` as elevating a tape transform, which is simply a "tape(s) goes in, tape(s) goes out" transform.

This is a case where I think it really makes sense to have a quantum function transform, in that we have a transform that takes a quantum function as input, and returns a _function that computes the matrix_ as output. So, something like:

```python
def qfunc(x, y, z):
    qml.RX(x, wires=0)
    qml.RY(y + z, wires=0)
    return qml.expval(qml.PauliZ(wires=0))

mat_func = qml.transforms.get_unitary_matrix(qfunc)

x, y, z = 1, 2, 3
mat_func(x, y, z) # returns a matrix
```

But it is not a _quantum_ function that gets output. Looping in @josh146 since he's been working a lot on transforms, in case he has any other thoughts: this idea makes total sense as a QNode transform, but could it still be considered a quantum function transform? Is it maybe a different "class" of them than existing ones?

Reply:
Alternatively, it could also be written as a function that acts on _QNodes_, like `qml.draw()`:
```
def get_unitary_matrix(qnode):
    def wrapper(*args, **kwargs):
        qnode.construct(args, kwargs)
        tape = qnode.qtape        
        mat = np.zeros(...)

        for op in tape.operations:
            # accumulate matrices here
            mat = ...

        return mat
    return wrapper
```

Which makes more sense?

Reply:
Right. So, what would be the best way to deal with tape? When inputting a QNode, would it be ok to ignore the existing tape and just do `tape = qml.transforms.make_tape(qfunc)(*args, **kwargs)` in the function that takes qfunc as input?

Reply:
@ingstra you could do this, in order to support everything:

```python
def get_unitary_matrix(fn):

    def wrapper(*args, **kwargs):

        if isinstance(fn, qml.QNode):
            # user passed a QNode, get the tape
            fn.construct(args, kwargs)
            tape = fn.qtape

        elif isinstance(fn, qml.tape.QuantumTape):
            # user passed a tape
            tape = fn

        elif callable(fn):
            # user passed something that is callable but not a tape or qnode.
            # we'll assume it is a qfunc!
            tape = qml.transforms.make_tape(fn)(*args, **kwargs)

        mat = np.zeros(...)

        for op in tape.operations:
            # accumulate matrices here
            mat = ...

        return mat
    return wrapper
```